src/required_resources.h

```diff
@@ -103,6 +103,10 @@ class RequiredResources {
   // rather than duplicating it in CodegenResources.
   uint64_t max_fmtstring_args_size = 0;
 
+  // Required for sizing of tuple scratch buffer
+  size_t max_tuple_size = 0;
+  size_t max_tuple_depth = 0;
+
```

Comment:
Better fit for codegen_resources.h? Since you are using initial value actual_size trick, so these values do not need to be serialized for AOT

Reply:
Can't use codegen_resources.h because I need to set `needed_global_variables` based on these values.

I suppose I could introduce a dependency between codegen_resources and required_resources where required_resources checks codegen_resources and adds necessary needed_global_variables. What do you think about that or leaving it as it is?

Reply:
Sorry, not quite following. The only references I see to `max_tuple_size` and `max_tuple_depth` are from `globalvars::get_type()`. But `get_type()` is called like:

```
CodegenLLVM::generate_ir()
  CodegenLLVM::generate_global_vars()
    bpftrace::globalvars::get_type()
```

`CodegenResources` is present as a local variable in `CodegenLLVM::generate_ir()`, so it's possible to thread both `RequiredResources` and `CodegenResources` through to `get_type()` just like we do for `CodegenLLVM::generate_maps()` right?

Reply:
See this code in ResourceAnalyser::analyse() in my PR:
```
  if (resources_.max_tuple_size > 0) {
    assert(resources_.max_tuple_depth > 0);
    resources_.needed_global_vars.insert(
        bpftrace::globalvars::GlobalVar::TUPLE_BUFFER);
    resources_.needed_global_vars.insert(
        bpftrace::globalvars::GlobalVar::MAX_CPU_ID);
  }
```

We need to add MAX_CPU_ID and TUPLE_BUFFER to needed_global_vars for AOT to let AOT know to resize these read-write .data.* ELF regions.

Yea I suppose we could append to bpftrace.resources in CodegenLLVM::generate_global_vars() based on codegen_resoureces...but it feels a bit unclean to me as its introducing a dependency between those objects buried in CodegenLLVM module.

Reply:
Ah, got it. Missed the `> 0` checks. Let's leave it as-is then.